CalibMuon/DTCalibration/plugins/DTResidualHistory.cc

```diff
@@ -0,0 +1,308 @@
+#include "TH2F.h"
+
+#include <algorithm>
+
+using namespace std;
+
+DTResidualHistory::DTResidualHistory(const edm::ParameterSet& pset):
+  select_(pset),
+  segment4DLabel_(pset.getParameter<edm::InputTag>("segment4DLabel")),
+  rootBaseDir_(pset.getUntrackedParameter<std::string>("rootBaseDir","DT/Residuals")),
+  detailedAnalysis_(pset.getUntrackedParameter<bool>("detailedAnalysis",false)) {
+
+  edm::LogVerbatim("Calibration") << "[DTResidualHistory] Constructor called.";
+  consumes< DTRecSegment4DCollection >(edm::InputTag(segment4DLabel_));
+  std::string rootFileName = pset.getUntrackedParameter<std::string>("rootFileName","residuals.root");
+  rootFile_ = new TFile(rootFileName.c_str(), "RECREATE");
+  rootFile_->cd();
+  
+  segmok=0;
+  segmbad=0;
+  nevent=0;
+}
+
+DTResidualHistory::~DTResidualHistory() {
+  edm::LogVerbatim("Calibration") << "[DTResidualHistory] Destructor called.";
+  cout << endl << " Finished. " << endl;
+  cout << " Analyzed events: " << nevent << endl;
+  cout << " Good segments: " << segmok << endl;
+  cout << " Bad segments: " << segmbad << endl;
+}
+
+void DTResidualHistory::beginJob() {
+//  TH1::SetDefaultSumw2(true);
+}
+
+void DTResidualHistory::beginRun(const edm::Run& run, const edm::EventSetup& setup) {
+  
+  // get the geometry
+  edm::ESHandle<DTGeometry> dtGeomH; 
+  setup.get<MuonGeometryRecord>().get(dtGeomH);
+  dtGeom_ = dtGeomH.product();
+  lastrun = 0;
+}
+
+void DTResidualHistory::analyze(const edm::Event& event, const edm::EventSetup& setup) {
+
+  const DTSuperLayerId slIdToFill(-2,1,1,1);
+
+  rootFile_->cd();
+  ++nevent;
+  unsigned int run = event.id().run();
+  if (run!=lastrun) {
+     TH1AddDirectorySentry addDir;
+     histoMapTH1F_.clear();
+     auto ch_it = dtGeom_->chambers().begin(); 	 
+     auto ch_end = dtGeom_->chambers().end(); 	 
+     for (; ch_it != ch_end; ++ch_it) { 	 
+        std::vector<const DTSuperLayer*>::const_iterator sl_it = (*ch_it)->superLayers().begin(); 	 
+        std::vector<const DTSuperLayer*>::const_iterator sl_end = (*ch_it)->superLayers().end(); 	 
+        // Loop over the SLs 	 
+        for(; sl_it != sl_end; ++sl_it) { 
+           DTSuperLayerId slId = (*sl_it)->id();
+           bookHistos(slId,run);
+           if(detailedAnalysis_) {
+	      std::vector<const DTLayer*>::const_iterator layer_it = (*sl_it)->layers().begin(); 	 
+	      std::vector<const DTLayer*>::const_iterator layer_end = (*sl_it)->layers().end();
+	      for(; layer_it != layer_end; ++layer_it) { 
+		 DTLayerId layerId = (*layer_it)->id();
+		 bookHistos(layerId,run);
+              }
+           }
+        }
+      }    
+
+    std::stringstream runStr; runStr << run;
+    TDirectory* baseDir = rootFile_->GetDirectory(("Run"+runStr.str()).c_str());
+    if(!baseDir) baseDir = rootFile_->mkdir(("Run"+runStr.str()).c_str());
+    baseDir->cd();
+    histoResLs = new TH2F("histoResLs","Residuals vs Lumisection",100,0,10000,100,-1,1);
+    std::vector<TH2F*> histosTH2F;
```

Comment:
please clean up pointers here and elsewhere- probably the best solution is just to make them smart pointers

Reply:
to clarify - root does not "take care" of the memory for you. These need to be managed properly by your code.

Reply:
I believe in this case smart pointers won't work, in fact putting them in makes the code segfault at the end of the execution. This is because with the call to "TH1AddDirectorySentry addDir" all the histograms are being managed by the TFile and the call to TFile::Close() at the end deletes them. Then they try to delete themselves again when DTResidualHistory is destroyed